lib/readyset.rb

```diff
@@ -93,8 +91,43 @@ def self.drop_cache!(name_or_id: nil, sql: nil)
   #
   # @param [Array<Object>] *sql_array the SQL array to be executed against ReadySet
   # @return [PG::Result]
-  def self.raw_query(*sql_array)
-    ActiveRecord::Base.establish_connection(Readyset::Configuration.configuration.database_url)
-    ActiveRecord::Base.connection.execute(ActiveRecord::Base.sanitize_sql_array(sql_array))
+  def self.raw_query(*sql_array) # :nodoc:
+    ActiveRecord::Base.connected_to(role: reading_role, shard: shard, prevent_writes: false) do
+      ActiveRecord::Base.connection.execute(ActiveRecord::Base.sanitize_sql_array(sql_array))
+    end
+  end
+
+  # Routes to ReadySet any queries that occur in the given block. If `prevent_writes` is false, an
+  # attempt to execute a write within the given block will raise an error. Keep in mind that if
+  # `prevent_writes` is true, any writes that occur within the given block will be proxied through
+  # ReadySet to the database.
+  #
+  # @param [Boolean] prevent_writes prevent writes from being executed on the connection to ReadySet
+  # @yield a block whose queries should be routed to ReadySet
+  # @return the value of the last line of the block
+  def self.route(prevent_writes: true, &block)
+    if prevent_writes
+      ActiveRecord::Base.
+        connected_to(role: reading_role, shard: shard, prevent_writes: true, &block)
+    else
+      ActiveRecord::Base.
+        connected_to(role: writing_role, shard: shard, prevent_writes: false, &block)
+    end
+  end
+
+  private
+
+  class << self
+    private(*delegate(:shard, to: :configuration))
+  end
+
+  def self.reading_role
+    ActiveRecord.reading_role
+  end
+  private_class_method :reading_role
```

Comment:
Where is the `reading_role` being sourced from in reference to the `connection_url`?

Reply:
The reading role is unrelated to the `connection_url` and is set globally across all the different databases. By default, `ActiveRecord.reading_role` is `:reading` and `ActiveRecord.writing_role` is `:writing`, but Rails optionally provides a way for users to specify their own names for these roles via
```ruby
config.active_record.writing_role = :writing
config.active_record.reading_role = :reading
```
The docs [here](https://guides.rubyonrails.org/active_record_multiple_databases.html#setting-up-your-application) have some more info about this. I'm using `ActiveRecord.reading_role` and `ActiveRecord.writing_role` here instead of hard coding `:reading` and `:writing` to account for the fact that users may have configured custom names for these roles